src/main/java/sc/iview/SciView.java

```diff
@@ -1101,48 +1101,44 @@ public void setColormap( Node n, ColorTable colorTable ) {
         return v;
     }
 
-    public <T extends RealType<T>> graphics.scenery.Node updateVolume( IterableInterval<T> image, String name,
-                                                                       float[] voxelDimensions, Volume v ) {
+    public < T extends RealType< T > > Node updateVolume( RandomAccessibleInterval< T > image, String name,
+                                                          float[] voxelDimensions, Volume v ) {
         log.debug( "Update Volume" );
 
         long dimensions[] = new long[3];
         image.dimensions( dimensions );
 
-        @SuppressWarnings("unchecked") Class<T> voxelType = ( Class<T> ) image.firstElement().getClass();
-        int bytesPerVoxel = image.firstElement().getBitsPerPixel() / 8;
-        NativeTypeEnum nType;
-
-        if( voxelType == UnsignedByteType.class ) {
+        @SuppressWarnings("unchecked") T type = ( T ) Util.getTypeFromInterval( image );
+        final NativeTypeEnum nType;
+        if( type instanceof UnsignedByteType ) {
             nType = NativeTypeEnum.UnsignedByte;
-        } else if( voxelType == UnsignedShortType.class ) {
+        } else if( type instanceof UnsignedShortType ) {
             nType = NativeTypeEnum.UnsignedShort;
-        } else if( voxelType == FloatType.class ) {
+        } else if( type instanceof FloatType ) {
             nType = NativeTypeEnum.Float;
         } else {
-            log.debug( "Type: " + voxelType +
+            log.debug( "Type: " + type.getClass() +
                        " cannot be displayed as a volume. Convert to UnsignedByteType, UnsignedShortType, or FloatType." );
             return null;
         }
 
         // Make and populate a ByteBuffer with the content of the Dataset
-        ByteBuffer byteBuffer = ByteBuffer.allocateDirect(
-                ( int ) ( bytesPerVoxel * dimensions[0] * dimensions[1] * dimensions[2] ) );
-        Cursor<T> cursor = image.cursor();
-
-        while( cursor.hasNext() ) {
-            cursor.fwd();
-            if( voxelType == UnsignedByteType.class ) {
-                byteBuffer.put( ( byte ) ( ( ( UnsignedByteType ) cursor.get() ).get() ) );
-            } else if( voxelType == UnsignedShortType.class ) {
-                byteBuffer.putShort( ( short ) Math.abs( ( ( UnsignedShortType ) cursor.get() ).getShort() ) );
-            } else if( voxelType == FloatType.class ) {
-                byteBuffer.putFloat( ( ( FloatType ) cursor.get() ).get() );
-            }
+        final ByteBuffer byteBuffer = ByteBuffer.allocateDirect(
+                ( int ) ( nType.getSizeInBytes() * Intervals.numElements( dimensions ) ) );
```

Comment:
There is `org.scijava.util.ArrayUtils.safeMultiply32` to do a product of longs and guarantee the result fits into an `int`—or else an exception is thrown. Should we use it here?

Reply:
Sure, I'll add it.